utility_analysis_new/combiners.py

```diff
@@ -356,6 +359,67 @@ def metrics_names(self) -> List[str]:
         ]
 
 
+class CompoundCombiner(pipeline_dp.combiners.CompoundCombiner):
+    """Compound combiner for Utility analysis per partition metrics."""
+
+    # For improving memory usage the cacompound accumulator has 2 modes:
+    # 1. Sparse mode (for small datasets): which contains information about each
+    # privacy id contributions.
+    # 2. Dense mode (for large datasets): which contains underlying accumulators
+    # from internal combiners.
+    # Since the utility analysis can be run for many configurations, there can
+    # be hundreds of the internal combiners, as a result the compound
+    # accumulator can contain hundreds accumulators. Converting each privacy id
+    # contribution to such accumulators leads to memory usage blow-up. That is
+    # why sparse mode introduced - until the number of privacy id contributions
+    # is small, they are saved instead of creating accumulators.
+    SparseAccumulatorType = List[Tuple[int, float, int]]
+    DenseAccumulatorType = List[Any]
+    AccumulatorType = Tuple[Optional[SparseAccumulatorType],
+                            Optional[DenseAccumulatorType]]
+
+    def create_accumulator(self, data: Tuple[Sequence, int]) -> AccumulatorType:
+        if not data:
+            return ((0, 0, 0), None)
+
+        values, n_partitions = data
+        sum_ = sum((v for v in values if v is not None))
+        return ([(len(values), sum_, n_partitions)], None)
+
+    def _to_dense(self,
+                  sparse_acc: SparseAccumulatorType) -> DenseAccumulatorType:
+        result = None
+        for count_sum_n_partitions in sparse_acc:
+            compound_acc = (1, [
+                combiner.create_accumulator(count_sum_n_partitions)
+                for combiner in self._combiners
+            ])
+            if result is None:
+                result = compound_acc
+            else:
+                result = super().merge_accumulators(result, compound_acc)
+        return result
+
+    def merge_accumulators(self, acc1: AccumulatorType, acc2: AccumulatorType):
+        sparse1, dense1 = acc1
+        sparse2, dense2 = acc2
+        if sparse1 and sparse2:
+            sparse1.extend(sparse2)
+            if len(sparse1) <= 5 * len(self._combiners):
```

Comment:
How did we pick this `5` number?

(Also, it seems like a prime candidate for a constant)

Reply:
I've changed to `2`, it's approximate ratio, 1 privacy id aggregated ratio per accumulator size.
Yes, of course that's the candidate to const, so I've know idea how to call this const :). I've added explanation in comments. PTAL